File: web/js/rocket.js
class Rocket {
  constructor(x, y) {
    this.x = x;
    this.y = y;
    this.vx = 0;
    this.vy = 0;
    this.angle = -90; // Наверх
    this.radius = 15;
    this.fuel = CONFIG.INITIAL_FUEL;
    this.credits = CONFIG.INITIAL_CREDITS;
    this.cargo = { metal: 0, fuel: 0, crystals: 0, gas: 0 };
    this.thrusting = false;
    this.rotationLeft = false;
    this.rotationRight = false;
    this.landed = false; // Флаг посадки
    this.landedOnPlanet = null; // На какой планете сидим
  }

  update(dt) {
    // Если на планете - ничего не делаем пока не оторвёмся
    if (this.landed) {
      // Можно оторваться только с топливом
      if (this.thrusting && this.fuel > 0) {
        this.landed = false;
        this.landedOnPlanet = null;
      } else {
        return [];
      }
    }

    // Поворот
    if (this.rotationLeft) {
      this.angle -= CONFIG.ROTATION_SPEED * dt;
    }
    if (this.rotationRight) {
      this.angle += CONFIG.ROTATION_SPEED * dt;
    }

    // Тяга (только если есть топливо)
    if (this.thrusting && this.fuel > 0) {
      const angleRad = (this.angle * Math.PI) / 180;
      const thrustX = Math.cos(angleRad) * CONFIG.THRUST_POWER * dt;
      const thrustY = Math.sin(angleRad) * CONFIG.THRUST_POWER * dt;

      this.vx += thrustX;
      this.vy += thrustY;
      this.fuel -= CONFIG.FUEL_CONSUMPTION * dt;

      // Не даём уйти в минус
      if (this.fuel < 0) this.fuel = 0;

      // Создаём частицы огня
      return this.createThrustParticles();
    }

    // Обновление позиции
    this.x += this.vx * dt;
    this.y += this.vy * dt;

    // Границы экрана (зацикливание)
    if (this.x < 0) this.x = canvas.width;
    if (this.x > canvas.width) this.x = 0;
    if (this.y < 0) this.y = canvas.height;
    if (this.y > canvas.height) this.y = 0;

    return [];
  }

  createThrustParticles() {
    const angleRad = ((this.angle + 180) * Math.PI) / 180;
    const particles = [];

    for (let i = 0; i < 3; i++) {
      particles.push({
        x: this.x,
        y: this.y,
        vx: Math.cos(angleRad) * (100 + Math.random() * 50),
        vy: Math.sin(angleRad) * (100 + Math.random() * 50),
        life: 0.5,
        maxLife: 0.5,
        color: `hsl(${20 + Math.random() * 30}, 100%, 50%)`,
      });
    }

    return particles;
  }

  distanceTo(planet) {
    const dx = this.x - planet.x;
    const dy = this.y - planet.y;
    return Math.sqrt(dx * dx + dy * dy);
  }

  landOn(planet) {
    const angle = Math.atan2(planet.y - this.y, planet.x - this.x);
    this.x = planet.x - Math.cos(angle) * (planet.radius + 25);
    this.y = planet.y - Math.sin(angle) * (planet.radius + 25);
    this.vx = 0;
    this.vy = 0;
    this.thrusting = false;
    this.landed = true;
    this.landedOnPlanet = planet;
  }

  draw(ctx) {
    const angleRad = (this.angle * Math.PI) / 180;
    const rocketLength = 35;

    // Нос ракеты
    const noseX = this.x + Math.cos(angleRad) * rocketLength;
    const noseY = this.y + Math.sin(angleRad) * rocketLength;

    // Огонь двигателя
    if (this.thrusting && this.fuel > 0) {
      const flameLength = 20 + Math.random() * 10;
      const flameX = this.x - Math.cos(angleRad) * flameLength;
      const flameY = this.y - Math.sin(angleRad) * flameLength;

      const gradient = ctx.createLinearGradient(this.x, this.y, flameX, flameY);
      gradient.addColorStop(0, "#ff6600");
      gradient.addColorStop(1, "transparent");

      ctx.strokeStyle = gradient;
      ctx.lineWidth = 8;
      ctx.beginPath();
      ctx.moveTo(this.x, this.y);
      ctx.lineTo(flameX, flameY);
      ctx.stroke();
    }

    // Корпус ракеты
    ctx.strokeStyle = "#ffffff";
    ctx.lineWidth = 10;
    ctx.lineCap = "round";
    ctx.beginPath();
    ctx.moveTo(this.x, this.y);
    ctx.lineTo(noseX, noseY);
    ctx.stroke();

    // Нос ракеты (красный)
    ctx.fillStyle = "#ff3333";
    ctx.beginPath();
    ctx.arc(noseX, noseY, 5, 0, Math.PI * 2);
    ctx.fill();

    // Индикатор направления
    ctx.strokeStyle = "#00ffff";
    ctx.lineWidth = 2;
    ctx.beginPath();
    ctx.moveTo(this.x, this.y);
    ctx.lineTo(
      this.x + Math.cos(angleRad) * 50,
      this.y + Math.sin(angleRad) * 50,
    );
    ctx.stroke();
  }
}

File: src/entities/rocket.py
"""
Ракета - основная сущность игрока
"""

import pygame
import math
from src.config import Config


class Rocket:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.vx = 0
        self.vy = 0
        self.angle = -90  # Наверх
        self.thrusting = False
        self.radius = 10
        
        # Ресурсы ракеты
        self.fuel = 200
        self.cargo = {"metal": 0, "fuel": 0, "crystals": 0, "gas": 0}
        self.credits = Config.INITIAL_CREDITS
        
        # Параметры
        self.max_fuel = 500
        
        # Визуал
        self.color = (255, 255, 255)
        
    def update(self, dt):
        if self.thrusting and self.fuel > 0:
            # Ускорение в направлении ракеты
            thrust_x = math.cos(math.radians(self.angle)) * Config.THRUST_POWER * dt
            thrust_y = math.sin(math.radians(self.angle)) * Config.THRUST_POWER * dt
            
            self.vx += thrust_x
            self.vy += thrust_y
            self.fuel -= 50 * dt  # Расход топлива
            
        # Обновление позиции
        self.x += self.vx * dt
        self.y += self.vy * dt
        
    def start_thrust(self):
        self.thrusting = True
        
    def stop_thrust(self):
        self.thrusting = False
        
    def rotate(self, direction, dt):
        """direction: 1 или -1"""
        self.angle += Config.ROTATION_SPEED * direction * dt
        
    def land_on(self, planet):
        """Посадка на планету"""
        self.x = planet.x
        self.y = planet.y - planet.radius - 20
        self.vx = 0
        self.vy = 0
        self.thrusting = False
        
    def distance_to(self, planet):
        """Расстояние до планеты"""
        dx = self.x - planet.x
        dy = self.y - planet.y
        return math.sqrt(dx**2 + dy**2)
        
    def draw(self, screen):
        """Отрисовка ракеты"""
        # Корпус ракеты
        rocket_length = 30
        rocket_width = 10
        
        # Рассчитываем точки ракеты
        nose_x = self.x + math.cos(math.radians(self.angle)) * rocket_length
        nose_y = self.y + math.sin(math.radians(self.angle)) * rocket_length
        
        # Огонь двигателя
        if self.thrusting and self.fuel > 0:
            flame_length = 15 + 5 * abs(math.sin(pygame.time.get_ticks() / 50))
            flame_x = self.x - math.cos(math.radians(self.angle)) * flame_length
            flame_y = self.y - math.sin(math.radians(self.angle)) * flame_length
            
            pygame.draw.line(
                screen,
                (255, 165, 0),
                (self.x, self.y),
                (flame_x, flame_y),
                4
            )
        
        # Линия корпуса
        pygame.draw.line(
            screen,
            self.color,
            (self.x, self.y),
            (nose_x, nose_y),
            rocket_width
        )
        
        # Нос ракеты
        pygame.draw.circle(
            screen,
            (255, 0, 0),
            (int(nose_x), int(nose_y)),
            rocket_width // 2
        )
